File: scripts/client_script.py
import argparse
import paramiko
import json
import os
import select
import time
import re  # для работы с регулярными выражениями

def run_remote_command(command, client, timeout=2000):
	stdin, stdout, stderr = client.exec_command(command, get_pty=True)
	out_lines = []
	err_lines = []
	start_time = time.time()

	while True:
		# Используем select для ожидания данных в stdout или stderr с таймаутом
		if stdout.channel.exit_status_ready() and not stdout.channel.recv_ready() and not stderr.channel.recv_stderr_ready():
			break

		r, _, _ = select.select([stdout.channel], [], [], 1)
		if r:
			output = stdout.channel.recv(1024).decode("utf-8")
			if output:
				print(output, end="")
				out_lines.append(output)
				start_time = time.time()  # сброс таймаута при получении данных
		else:
			# Если данные не получены, проверяем таймаут
			if time.time() - start_time > timeout:
				print("\nTimeout reached, assuming command completed.")
				break

	remaining_output = ""
	while stdout.channel.recv_ready():
		chunk = stdout.channel.recv(1024).decode("utf-8")
		if not chunk:
			break
		remaining_output += chunk

	remaining_error = ""
	while stderr.channel.recv_stderr_ready():
		chunk = stderr.channel.recv_stderr(1024).decode("utf-8")
		if not chunk:
			break
		remaining_error += chunk

	if remaining_output:
		print(remaining_output, end="")
		out_lines.append(remaining_output)
	if remaining_error:
		print(remaining_error, end="")
		err_lines.append(remaining_error)

	return "".join(out_lines).strip() if out_lines else None

# Функция подключения к серверу
def connect_to_server(hostname, port, username, password):
	try:
		client = paramiko.SSHClient()
		client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
		client.connect(hostname, port=int(port), username=username, password=password)
		print(f"Connected to server {hostname}.")
		return client
	except Exception as e:
		print(f"Connection error: {e}")
		return None

# Функция создания конфига для сервера
def create_server_config(client, user_port, server_names):
	uuid = run_remote_command("/usr/local/bin/xray uuid", client)
	short_id = run_remote_command("openssl rand -hex 8", client)
	x25519_keys = run_remote_command("/usr/local/bin/xray x25519", client)

	if not all([uuid, short_id, x25519_keys]):
		print("Error: Failed to obtain all necessary data from the server.")
		return None

	private_key, public_key = "", ""
	for line in x25519_keys.split("\n"):
		if "Private" in line:
			private_key = line.split(":")[-1].strip()
		if "Public" in line:
			public_key = line.split(":")[-1].strip()

	dest = f"{server_names.split(',')[0]}:{user_port}"

	return {
		"log": {"loglevel": "info"},
		"routing": {"rules": [], "domainStrategy": "AsIs"},
		"inbounds": [
			{
				"port": 23,
				"tag": "ss",
				"protocol": "shadowsocks",
				"settings": {
					"method": "2022-blake3-aes-128-gcm",
					"password": "aaaaaaaaaaaaaaaabbbbbbbbbbbbbbbb",
					"network": "tcp,udp"
				}
			},
			{
				"port": int(user_port),
				"protocol": "vless",
				"tag": "vless_tls",
				"settings": {
					"clients": [{"id": uuid, "email": "user1@myserver", "flow": "xtls-rprx-vision"}],
					"decryption": "none"
				},
				"streamSettings": {
					"network": "tcp",
					"security": "reality",
					"realitySettings": {
						"show": False,
						"dest": dest,
						"xver": 0,
						"serverNames": server_names.split(','),
						"privateKey": private_key,
						"minClientVer": "",
						"maxClientVer": "",
						"maxTimeDiff": 0,
						"shortIds": [short_id]
					}
				},
				"sniffing": {"enabled": True, "destOverride": ["http", "tls"]}
			}
		],
		"outbounds": [
			{"protocol": "freedom", "tag": "direct"},
			{"protocol": "blackhole", "tag": "block"}
		]
	}, uuid, private_key, short_id, dest, public_key

# Функция загрузки конфига на сервер
def upload_config_to_server(client, server_config, uuid, private_key, short_id, dest, public_key):
	try:
		remote_config_path = "/usr/local/etc/xray/config.json"
		sftp = client.open_sftp()
		with sftp.file(remote_config_path, "w") as remote_file:
			remote_file.write(json.dumps(server_config, indent=2))
		sftp.close()
		run_remote_command("systemctl restart xray", client)

		print(f"Configuration successfully uploaded to the server and saved at {remote_config_path}")
		print("UUID:", uuid)
		print("Private Key:", private_key)
		print("Short ID:", short_id)
		print("Dest:", dest)
		print("Public Key:", public_key)

	except Exception as e:
		print(f"Error uploading config to server: {e}")

# Функция создания клиентского конфига
def create_client_config(hostname, server_names, uuid, public_key, short_id):
	return {
		"log": {"level": "info"},
		"inbounds": [
			{
				"type": "mixed",
				"tag": "mixed-in",
				"listen": "127.0.0.1",
				"listen_port": 2080,
				"tcp_fast_open": False,
				"sniff": True,
				"sniff_override_destination": True,
				"set_system_proxy": False
			}
		],
		"outbounds": [
			{
				"type": "vless",
				"tag": "vless-out",
				"server": hostname,
				"server_port": 443,
				"uuid": uuid,
				"flow": "xtls-rprx-vision",
				"network": "tcp",
				"tls": {
					"enabled": True,
					"server_name": server_names.split(',')[0],
					"alpn": ["h2"],
					"utls": {
						"enabled": True,
						"fingerprint": "chrome"
					},
					"reality": {
						"enabled": True,
						"public_key": public_key,
						"short_id": short_id
					}
				}
			}
		]
	}

# Функция сохранения клиентского конфига в файл с заданным именем
def save_client_config(local_folder, client_config, filename):
	if not os.path.exists(local_folder):
		try:
			os.makedirs(local_folder)
			print(f"Folder {local_folder} created.")
		except Exception as e:
			print(f"Error creating folder: {e}")
			return

	if not filename.endswith(".json"):
		filename += ".json"
	local_config_path = os.path.join(local_folder, filename)

	try:
		with open(local_config_path, "w") as file:
			json.dump(client_config, file, indent=2)
		print(f"Configuration file saved: {local_config_path}")
	except Exception as e:
		print(f"Error saving file: {e}")

def main():
	parser = argparse.ArgumentParser(description="Utility for configuring Xray on the server")
	parser.add_argument("-ip", "--ipaddress", required=True, help="Server IP address")
	parser.add_argument("-p", "--password", required=True, help="Server password")
	parser.add_argument("--port", default=22, help="Server SSH port (default 22)")
	parser.add_argument("-uport", "--user_port", required=True, help="Port for creating the server configuration")
	parser.add_argument("-s", "--server_names", required=True, help="Comma-separated list of serverNames (e.g., www.example.com,www.google.com)")
	parser.add_argument("-l", "--local_folder", required=True, help="Path to folder for saving the client config")
	parser.add_argument("-f", "--filename", required=True, help="Filename for saving the client config (without extension)")
	
	args = parser.parse_args()

	hostname = args.ipaddress
	port = args.port
	username = "root"
	password = args.password

	client = connect_to_server(hostname, port, username, password)
	if not client:
		return

	# Проверка установки Xray
	print("Checking Xray installation...")
	xray_check = run_remote_command("command -v xray", client)
	if xray_check:
		print("Xray is already installed, skipping installation.")
	else:
		print("Xray is not installed. Proceeding with installation...")
		print("Running command: apt upgrade")
		update_output = run_remote_command("apt upgrade -y", client)
		if update_output is not None:
			# Если в выводе присутствует сообщение об ожидании блокировки
			if "Waiting for cache lock:" in update_output:
				match = re.search(r"lock-frontend\. It is held by process (\d+)", update_output)
				if match:
					pid = match.group(1)
					print(f"Detected apt-get process with PID {pid}. Attempting to kill it...")
					kill_output = run_remote_command(f"sudo kill -9 {pid}", client)
					print("apt-get process killed.")
				else:
					print("Cache lock message detected, but failed to parse process ID.")
			else:
				print("apt upgrade executed successfully.")
		else:
			print("Error executing apt upgrade.")
		
		install_command = 'bash -c "$(curl -L https://raw.githubusercontent.com/XTLS/Xray-install/046d9aa2432b3a6241d73c3684ef4e512974b594/install-release.sh)" @ install --version 1.8.24'
		print("Installing Xray using the installation script...")
		install_output = run_remote_command(install_command, client)
		if install_output is not None:
			print("Xray installation script executed successfully.")
		else:
			print("Error executing Xray installation script.")

	user_port = args.user_port
	server_names = args.server_names

	result = create_server_config(client, user_port, server_names)
	if not result:
		client.close()
		return

	server_config, uuid, private_key, short_id, dest, public_key = result
	upload_config_to_server(client, server_config, uuid, private_key, short_id, dest, public_key)

	client_config = create_client_config(hostname, server_names, uuid, public_key, short_id)
	save_client_config(args.local_folder, client_config, args.filename)

	client.close()

if __name__ == "__main__":
	main()

File: scripts/client_script.go
package scripts

import (
	"bufio"
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
	"time"

	"golang.org/x/crypto/ssh"
)

// ServerConfig описывает конфигурацию для xray.
type ServerConfig struct {
	Log       map[string]string        `json:"log"`
	Routing   map[string]interface{}   `json:"routing"`
	Inbounds  []map[string]interface{} `json:"inbounds"`
	Outbounds []map[string]interface{} `json:"outbounds"`
}

// runRemoteCommand выполняет команду на сервере с заданным таймаутом через pty.
func runRemoteCommand(client *ssh.Client, command string, timeout time.Duration) (string, error) {
	session, err := client.NewSession()
	if err != nil {
		return "", err
	}
	defer session.Close()

	if err := session.RequestPty("xterm", 80, 40, ssh.TerminalModes{}); err != nil {
		return "", err
	}

	var outputBuf bytes.Buffer
	session.Stdout = &outputBuf
	session.Stderr = &outputBuf

	if err := session.Start(command); err != nil {
		return "", err
	}

	done := make(chan error, 1)
	go func() {
		done <- session.Wait()
	}()

	select {
	case err := <-done:
		return outputBuf.String(), err
	case <-time.After(timeout):
		_ = session.Signal(ssh.SIGKILL)
		return outputBuf.String() + "\nTimeout reached, assuming command completed.", nil
	}
}

// ConnectToServer устанавливает SSH-соединение.
func ConnectToServer(host string, port int, username, password string) (*ssh.Client, error) {
	config := &ssh.ClientConfig{
		User:            username,
		Auth:            []ssh.AuthMethod{ssh.Password(password)},
		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
		Timeout:         5 * time.Second,
	}
	addr := fmt.Sprintf("%s:%d", host, port)
	client, err := ssh.Dial("tcp", addr, config)
	if err != nil {
		return nil, err
	}
	fmt.Printf("Connected to server %s.\n", host)
	return client, nil
}

// runRemoteCommandNoPTY выполняет команду без выделения pty.
func runRemoteCommandNoPTY(client *ssh.Client, command string, timeout time.Duration) (string, error) {
	session, err := client.NewSession()
	if err != nil {
		return "", err
	}
	defer session.Close()

	output, err := session.CombinedOutput(command)
	return string(output), err
}

// runRemoteCommandStream выполняет команду и передаёт потоковый вывод через logFunc.
func runRemoteCommandStream(client *ssh.Client, command string, timeout time.Duration, logFunc func(string)) (string, error) {
	session, err := client.NewSession()
	if err != nil {
		return "", err
	}
	defer session.Close()

	if err := session.RequestPty("xterm", 80, 40, ssh.TerminalModes{}); err != nil {
		return "", err
	}

	stdoutPipe, err := session.StdoutPipe()
	if err != nil {
		return "", err
	}
	stderrPipe, err := session.StderrPipe()
	if err != nil {
		return "", err
	}

	if err := session.Start(command); err != nil {
		return "", err
	}

	outputBuf := new(bytes.Buffer)
	done := make(chan error, 1)

	go func() {
		multiR := io.MultiReader(stdoutPipe, stderrPipe)
		scanner := bufio.NewScanner(multiR)
		for scanner.Scan() {
			line := scanner.Text() + "\n"
			outputBuf.WriteString(line)
			if logFunc != nil {
				logFunc(line)
			}
		}
		done <- session.Wait()
	}()

	select {
	case err := <-done:
		return outputBuf.String(), err
	case <-time.After(timeout):
		_ = session.Signal(ssh.SIGKILL)
		return outputBuf.String() + "\nTimeout reached, assuming command completed.", nil
	}
}

// detectOS определяет тип операционной системы, выполняя "cat /etc/os-release".
func detectOS(client *ssh.Client, timeout time.Duration) (string, error) {
	output, err := runRemoteCommandNoPTY(client, "cat /etc/os-release", timeout)
	if err != nil {
		return "", err
	}
	// Если присутствует ID=arch - Arch Linux
	if strings.Contains(output, "ID=arch") {
		return "arch", nil
	}
	// Если присутствует ID=debian или ID=ubuntu - считаем это debian-подобной системой
	if strings.Contains(output, "ID=debian") || strings.Contains(output, "ID=ubuntu") {
		return "debian", nil
	}
	return "unknown", nil
}

// createServerConfig получает данные с сервера и формирует конфигурацию.
func createServerConfig(client *ssh.Client, userPort string, serverNames string) (ServerConfig, string, string, string, string, string, error) {
	timeout := 2 * time.Second

	uuid, err := runRemoteCommandNoPTY(client, "/usr/local/bin/xray uuid", timeout)
	if err != nil || strings.TrimSpace(uuid) == "" {
		return ServerConfig{}, "", "", "", "", "", fmt.Errorf("failed to obtain uuid")
	}
	uuid = strings.TrimSpace(uuid)

	shortID, err := runRemoteCommandNoPTY(client, "openssl rand -hex 8", timeout)
	if err != nil || strings.TrimSpace(shortID) == "" {
		return ServerConfig{}, "", "", "", "", "", fmt.Errorf("failed to obtain shortID")
	}
	shortID = strings.TrimSpace(shortID)

	x25519Keys, err := runRemoteCommandNoPTY(client, "/usr/local/bin/xray x25519", timeout)
	if err != nil || strings.TrimSpace(x25519Keys) == "" {
		return ServerConfig{}, "", "", "", "", "", fmt.Errorf("failed to obtain x25519 keys")
	}

	var privateKey, publicKey string
	scanner := bufio.NewScanner(strings.NewReader(x25519Keys))
	for scanner.Scan() {
		line := scanner.Text()
		if strings.Contains(line, "Private") {
			parts := strings.Split(line, ":")
			if len(parts) > 1 {
				privateKey = strings.TrimSpace(parts[len(parts)-1])
			}
		}
		if strings.Contains(line, "Public") {
			parts := strings.Split(line, ":")
			if len(parts) > 1 {
				publicKey = strings.TrimSpace(parts[len(parts)-1])
			}
		}
	}
	if privateKey == "" || publicKey == "" {
		return ServerConfig{}, "", "", "", "", "", fmt.Errorf("failed to parse x25519 keys")
	}

	dest := fmt.Sprintf("%s:%s", strings.Split(serverNames, ",")[0], userPort)

	serverConfig := ServerConfig{
		Log: map[string]string{
			"loglevel": "info",
		},
		Routing: map[string]interface{}{
			"rules":          []interface{}{},
			"domainStrategy": "AsIs",
		},
		Inbounds: []map[string]interface{}{
			{
				"port":     23,
				"tag":      "ss",
				"protocol": "shadowsocks",
				"settings": map[string]interface{}{
					"method":   "2022-blake3-aes-128-gcm",
					"password": "aaaaaaaaaaaaaaaabbbbbbbbbbbbbbbb",
					"network":  "tcp,udp",
				},
			},
			{
				"port":     stringToInt(userPort),
				"protocol": "vless",
				"tag":      "vless_tls",
				"settings": map[string]interface{}{
					"clients": []map[string]interface{}{
						{
							"id":    uuid,
							"email": "user1@myserver",
							"flow":  "xtls-rprx-vision",
						},
					},
					"decryption": "none",
				},
				"streamSettings": map[string]interface{}{
					"network":  "tcp",
					"security": "reality",
					"realitySettings": map[string]interface{}{
						"show":         false,
						"dest":         dest,
						"xver":         0,
						"serverNames":  strings.Split(serverNames, ","),
						"privateKey":   privateKey,
						"minClientVer": "",
						"maxClientVer": "",
						"maxTimeDiff":  0,
						"shortIds":     []string{shortID},
					},
				},
				"sniffing": map[string]interface{}{
					"enabled":      true,
					"destOverride": []string{"http", "tls"},
				},
			},
		},
		Outbounds: []map[string]interface{}{
			{
				"protocol": "freedom",
				"tag":      "direct",
			},
			{
				"protocol": "blackhole",
				"tag":      "block",
			},
		},
	}
	return serverConfig, uuid, privateKey, shortID, dest, publicKey, nil
}

func stringToInt(s string) int {
	i, err := strconv.Atoi(s)
	if err != nil {
		return 0
	}
	return i
}

// uploadConfigToServer загружает конфигурацию на сервер и перезапускает сервис.
func uploadConfigToServer(client *ssh.Client, serverConfig ServerConfig, uuid, privateKey, shortID, dest, publicKey string) error {
	configJSON, err := json.MarshalIndent(serverConfig, "", "  ")
	if err != nil {
		return err
	}
	remotePath := "/usr/local/etc/xray/config.json"

	// Передадим конфиг через временный файл.
	tmpFile := "temp_config.json"
	if err := os.WriteFile(tmpFile, configJSON, 0644); err != nil {
		return err
	}
	defer os.Remove(tmpFile)

	scpCmd := fmt.Sprintf("cat > %s", remotePath)
	session, err := client.NewSession()
	if err != nil {
		return err
	}
	defer session.Close()

	fileContent, err := os.ReadFile(tmpFile)
	if err != nil {
		return err
	}
	stdin, err := session.StdinPipe()
	if err != nil {
		return err
	}
	go func() {
		defer stdin.Close()
		io.Copy(stdin, bytes.NewReader(fileContent))
	}()

	if err := session.Run(scpCmd); err != nil {
		return err
	}
	_, err = runRemoteCommand(client, "systemctl restart xray", 2*time.Second)
	if err != nil {
		return err
	}
	fmt.Printf("Configuration successfully uploaded to %s\n", remotePath)
	fmt.Printf("UUID: %s\nPrivate Key: %s\nShort ID: %s\nDest: %s\nPublic Key: %s\n", uuid, privateKey, shortID, dest, publicKey)
	return nil
}

// createClientConfig формирует клиентскую конфигурацию в виде map.
func createClientConfig(hostname, serverNames, uuid, publicKey, shortID string) map[string]interface{} {
	return map[string]interface{}{
		"log": map[string]interface{}{
			"level": "info",
		},
		"inbounds": []map[string]interface{}{
			{
				"type":                       "mixed",
				"tag":                        "mixed-in",
				"listen":                     "127.0.0.1",
				"listen_port":                2080,
				"tcp_fast_open":              false,
				"sniff":                      true,
				"sniff_override_destination": true,
				"set_system_proxy":           false,
			},
		},
		"outbounds": []map[string]interface{}{
			{
				"type":        "vless",
				"tag":         "vless-out",
				"server":      hostname,
				"server_port": 443,
				"uuid":        uuid,
				"flow":        "xtls-rprx-vision",
				"network":     "tcp",
				"tls": map[string]interface{}{
					"enabled":     true,
					"server_name": strings.Split(serverNames, ",")[0],
					"alpn":        []string{"h2"},
					"utls": map[string]interface{}{
						"enabled":     true,
						"fingerprint": "chrome",
					},
					"reality": map[string]interface{}{
						"enabled":    true,
						"public_key": publicKey,
						"short_id":   shortID,
					},
				},
			},
		},
	}
}

// saveClientConfig сохраняет клиентский конфиг в указанный файл.
func saveClientConfig(localFolder string, clientConfig map[string]interface{}, filename string) error {
	if _, err := os.Stat(localFolder); os.IsNotExist(err) {
		if err := os.MkdirAll(localFolder, 0755); err != nil {
			return fmt.Errorf("Error creating folder: %v", err)
		}
		fmt.Printf("Folder %s created.\n", localFolder)
	}
	if !strings.HasSuffix(filename, ".json") {
		filename += ".json"
	}
	localConfigPath := localFolder + "/" + filename
	configJSON, err := json.MarshalIndent(clientConfig, "", "  ")
	if err != nil {
		return err
	}
	if err := os.WriteFile(localConfigPath, configJSON, 0644); err != nil {
		return err
	}
	fmt.Printf("Configuration file saved: %s\n", localConfigPath)
	return nil
}

// checkXrayInstallation проверяет наличие xray на сервере.
func checkXrayInstallation(client *ssh.Client) bool {
	output, err := runRemoteCommand(client, "command -v xray", 2*time.Second)
	if err != nil {
		return false
	}
	return strings.TrimSpace(output) != ""
}

// installXray устанавливает xray, если он не установлен, с учетом типа системы.
func installXray(client *ssh.Client, osType string, logFunc func(string)) error {
	if osType == "arch" {
		logFunc("Running pacman update...\n")
		output, err := runRemoteCommandStream(client, "pacman -Syu --noconfirm", 60*time.Second, logFunc)
		if err != nil {
			logFunc("error executing pacman update.\n")
		} else {
			if strings.Contains(strings.ToLower(output), "error") || strings.Contains(strings.ToLower(output), "failed") {
				logFunc("pacman update completed with errors. Please resolve them manually.\n")
			} else {
				logFunc("pacman update executed successfully.\n")
			}
		}

		// Устанавливаем зависимости для xray
		logFunc("Installing dependencies (unzip, wget, ca-certificates)...\n")
		depCmd := "pacman -S --noconfirm unzip wget ca-certificates"
		depOut, depErr := runRemoteCommandStream(client, depCmd, 60*time.Second, logFunc)
		if depErr != nil {
			logFunc("error installing dependencies: " + depOut)
			return fmt.Errorf("error installing dependencies: %v", depErr)
		} else {
			logFunc("Dependencies installed successfully.\n")
		}
	} else {
		logFunc("Running apt upgrade...\n")
		output, err := runRemoteCommandStream(client, "apt upgrade -y", 60*time.Second, logFunc)
		if err == nil {
			if strings.Contains(output, "Waiting for cache lock:") {
				logFunc("cache lock detected. Please resolve the issue manually.\n")
			} else {
				logFunc("apt upgrade executed successfully.\n")
			}
		} else {
			logFunc("error executing apt upgrade.\n")
		}
	}

	installCommand := `bash -c "$(curl -L https://raw.githubusercontent.com/XTLS/Xray-install/046d9aa2432b3a6241d73c3684ef4e512974b594/install-release.sh)" @ install`
	logFunc("Installing Xray using the installation script...\n")
	installOutput, err := runRemoteCommandStream(client, installCommand, 120*time.Second, logFunc)
	if err != nil || strings.TrimSpace(installOutput) == "" {
		logFunc("error executing xray installation script\n")
		return fmt.Errorf("error executing xray installation script")
	}
	logFunc("Xray installation executed successfully.\n")
	return nil
}

// ServerSetup объединяет все шаги настройки сервера. Он автоматически определяет операционную систему.
func ServerSetup(ip, pass, sshPort, userPort, serverNames, localFolder, fileName string, logFunc func(string)) error {
	port, _ := strconv.Atoi(sshPort)
	client, err := ConnectToServer(ip, port, "root", pass)
	if err != nil {
		return fmt.Errorf("connection error: %v", err)
	}
	defer client.Close()
	logFunc("Connected to server.\n")

	// Автоматическое определение ОС
	logFunc("Detecting operating system...\n")
	osType, err := detectOS(client, 5*time.Second)
	if err != nil {
		logFunc("Error detecting OS, defaulting to debian.\n")
		osType = "debian"
	} else {
		logFunc(fmt.Sprintf("Detected OS: %s\n", osType))
	}

	logFunc("Checking Xray installation...\n")
	if checkXrayInstallation(client) {
		logFunc("Xray is already installed, skipping installation.\n")
	} else {
		logFunc("Xray is not installed. Installing...\n")
		if err := installXray(client, osType, logFunc); err != nil {
			return fmt.Errorf("installation error: %v", err)
		}
	}

	logFunc("Creating server configuration...\n")
	serverConfig, uuid, privateKey, shortID, dest, publicKey, err := createServerConfig(client, userPort, serverNames)
	if err != nil {
		return fmt.Errorf("error creating server config: %v", err)
	}

	logFunc("Uploading configuration to server...\n")
	if err := uploadConfigToServer(client, serverConfig, uuid, privateKey, shortID, dest, publicKey); err != nil {
		return fmt.Errorf("error uploading config: %v", err)
	}

	clientConfig := createClientConfig(ip, serverNames, uuid, publicKey, shortID)
	logFunc("Saving client configuration locally...\n")
	if err := saveClientConfig(localFolder, clientConfig, fileName); err != nil {
		return fmt.Errorf("error saving client config: %v", err)
	}
	logFunc("Server setup completed successfully.\n")
	return nil
}
